src/net/server/handlers/query.rs

```diff
@@ -64,13 +65,36 @@ impl<T> Clone for ReservedPermit<T> {
 
 /// Middleware that first reserves a permit on the channel to send messages to the messaging layer.
 /// Once reserved, adds the permit to the extension for retrieval from the handler.
-pub async fn obtain_permit_mw<T: Send + 'static, B>(
-    sender: mpsc::Sender<T>,
-    mut req: Request<B>,
+/// # Panics
+/// if messages arrive out of order
+pub async fn obtain_permit_mw<B: Send>(
+    req: Request<B>,
     next: Next<B>,
 ) -> Result<Response, MpcHelperServerError> {
+    let mut req_parts = RequestParts::new(req);
+    let Path(query_id, step) = Path::from_request(&mut req_parts).await?;
```

Comment:
This seems to do exactly what Axum extractors do. In other words

```rust
pub async fn obtain_permit_mw<B: Send>(
    Path(query_id, step): Path(QueryId, Step),
)
```

is there a reason you want to parse everything inside the handler?

Reply:
extractors only work on handlers, [not middleware](https://docs.rs/axum/latest/axum/middleware/fn.from_fn.html)...

Reply:
that is sad :(